### PR TITLE
FOGL-3215: initial checkin

### PR DESCRIPTION
**Pull request by pintomax**
FOGL-3215: initial checkin

### CODE REVIEW

**Comment by pintomax** [comment]
Tested only in PiZero Buster:

reconfiguration works using sleep(), leds.on() / off()

Still checking some deadlocks during instance enable/disable

To be tested on Ubuntu 16 and Buster on Pi3

**Comment by pintomax** [comment]
It looks like the change also fixes FOGL-3200

File: notify_python35.cpp [hunk 1]
```diff
@@ -220,20 +213,20 @@ bool NotifyPython35::reconfigure(const std::string& newConfig)
...
-		m_pythonScript.clear();
+	PyGILState_STATE state = PyGILState_Ensure(); // acquire GIL
 
-		// Remove Interpreter
-		Py_Finalize();
-		// Start interpreter, setting paths etc
-		interpreterStart(this);
-	}
+	// Reimport module
+	PyObject* newModule = PyImport_ReloadModule(m_pModule);
```
Comment (AmandeepArora) [inline notify_python35.cpp]: Check return value of PyImport_ReloadModule. And if this call fails, should we clear m_pModule & m_pFunc?

File: notify_python35.cpp [hunk 2]
```diff
@@ -151,15 +147,12 @@ bool NotifyPython35::configure()
 		return true;
 	}
 
-	// Check the Python interpreter is set
-	if (!Py_IsInitialized())
+	// 2) Import Python script if module object is not set
+	if (!m_pModule)
 	{
-		interpreterStart(this);
+		m_pModule = PyImport_ImportModule(m_pythonScript.c_str());
```
Comment (AmandeepArora) [inline notify_python35.cpp]: I think we should acquire and hold GIL for remainder of this function.

File: notify_python35.cpp [hunk 3]
```diff
@@ -297,33 +296,22 @@ bool NotifyPython35::notify(const std::string& deliveryName,
 			    const string& triggerReason,
 			    const string& customMessage)
 {
+	//lock_guard<mutex> guard(m_configMutex);
```
Comment (AmandeepArora) [inline notify_python35.cpp]: We should probably hold this mutex here. We don't want this function to interleave with (re)configure. E.g. in reconfigure, we do "m_pFunc = NULL" and here we are using "m_pFunc".

File: notify_python35.cpp (hunk 2 above)
Comment (pintomax) [reply notify_python35.cpp]: NotifyPython35::configure()

is always called while holding the GIL

NotifyPython35::reconfigure
        ....
        PyGILState_STATE state = PyGILState_Ensure();
        ...
        bool ret = this->configure();

        PyGILState_Release(state);


and

PLUGIN_HANDLE plugin_init(ConfigCategory* config)

     PyGILState_STATE state = PyGILState_Ensure(); // acquire GIL
    ....
      // Configure plugin
      notify->lock();
      bool ret = notify->configure();
      notify->unlock();

      PyGILState_Release(state); // release GIL

Comment (AmandeepArora) [reply notify_python35.cpp]: It is better to acquire and release GIL where required. Try not to hold GIL across main functions. Might introduce errors during code changes in future.

Comment (pintomax) [reply notify_python35.cpp]: @AmandeepArora this plugin is now similar to foglamp-filter-python35, same base code for init, configure and reconfigure.

As per the comment we should modify both.

 It's not needed now the change so a new JIRA task could be raised for both plugins

File: include/notify_python35.h [hunk 4]
```diff
@@ -62,6 +62,8 @@ class NotifyPython35
 		PyObject*	m_pModule;
 		// Python 3.5 callable method handle
 		PyObject*	m_pFunc;
+		// Python interpreter has been started by this plugin
+		bool		m_init;
```
Comment (MarkRiddoch) [inline include/notify_python35.h]: This should be initialised in the constructor.

File: plugin.cpp [hunk 5]
```diff
@@ -111,21 +108,71 @@ PLUGIN_HANDLE plugin_init(ConfigCategory* config)
...
+
+	// Embedded Python 3.5 initialisation
+	// Check first the interpreter is already set
+	if (!Py_IsInitialized())
+	{
+#ifdef PLUGIN_PYTHON_SHARED_LIBRARY
+		string openLibrary = TO_STRING(PLUGIN_PYTHON_SHARED_LIBRARY);
+		if (!openLibrary.empty())
+		{
+			libpython_handle = dlopen(openLibrary.c_str(),
```
Comment (MarkRiddoch) [inline plugin.cpp]: It woudl be good to check dlopen's return here.

**Comment by pintomax** [comment]
@mshariq-nerd Still testing in Ubuntu. Try Buster first

**Comment by pintomax** [comment]
@mshariq-nerd last commit was after Ubuntu tests. You can continue now.

**Pull request merged**